In [3]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix


In [6]:
df=pd.read_csv('news.csv') #read data into dataframe

In [7]:
df.head() #check data

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [9]:
df.shape #check shape of data

(6335, 4)

In [11]:
df.isnull().sum() #check for null input/values

Unnamed: 0    0
title         0
text          0
label         0
dtype: int64

In [12]:
#get the target values from the dataset
labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [15]:
#split the data into training and testing test
x_train,x_test,y_train,y_test = train_test_split(df['text'], labels,test_size=0.2, random_state=7)

### TF-IDF is a statistical measure that evaluates how relevant a word is to a document in a collection of documents. This is done by multiplying two metrics: how many times a word appears in a document, and the inverse document frequency of the word across a set of documents.

### Initialize a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7 (terms with a higher document frequency will be discarded). Stop words are the most common words in a language that are to be filtered out before processing the natural language data. And a TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features.

In [18]:
tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_df=0.7) #initialize a TfidfVectorizer


In [19]:
#fit and transform train set, transform test set
tfidf_train = tfidf_vectorizer.fit_transform(x_train)
tfidf_test = tfidf_vectorizer.transform(x_test)

### Initialize a PassiveAggressiveClassifier, then fit it on tdidf_train and y_train 

Passive Aggressive are considered algorithms that perform online learning (with for example twitter data). Their characteristic is that they remain passive when dealing with an outcome which has been correctly classified, and become aggressive when a miscalculation takes place, thus constantly self-updating and adjusting.

In [24]:
model = PassiveAggressiveClassifier(max_iter =50)
model.fit(tfidf_train, y_train)

PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=50, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

### Predict on the test set (i.e tfidf_test)

In [29]:
y_pred = model.predict(tfidf_test)

### calculate the accuracy with accuracy_score from sklearn.metrics

In [30]:
score = accuracy_score(y_test, y_pred)

### Print Accuracy score

In [32]:
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.74%


### print out a confusion matrix to compares the actual target values with those predicted by the model.

In [33]:
confusion_matrix(y_test, y_pred, labels =['FAKE', 'REAL'])

array([[589,  49],
       [ 43, 586]], dtype=int64)

The above means we have 589 true positives, 587 true negatives, 42 false positives